### ISSUE TITLE
MLIR based backend for Numba

### ISSUE THREAD

**Issue by Hardcode84** [numba/numba#6475]
Just to share and discuss our ongoing work on MLIR based backend for Numba.

https://github.com/IntelPython/numba/pull/115

It is very limited at the moment but it already can run simple codes

**Comment by sklam** [comment]
@Hardcode84, I just briefly looked through the code and have a few questions:

- Is the plier_to_std.cpp recreating high-level control-flow (scf) from the gotos?
- Out of curiously, why is the dialect called Plier?

**Comment by Hardcode84** [comment]
> Is the plier_to_std.cpp recreating high-level control-flow (scf) from the gotos?

Yes, I convert branches to `scf.If` and loops to `scf.While`, and then `range` loops from `scf.While` to `scf.For`

> Out of curiously, why is the dialect called Plier?

CC @diptorupd

**Comment by diptorupd** [comment]
Because with this PLIER we will rip out any nails that can prick us in the future. On a more intellectual level it is an acronym correctly written as PL(IE)_t_R : Pythonic Languages Extensible Intermediate Representation dialect. (Note the transpose operator to flip the `I` and `E` ;).)

**Comment by animator** [comment]
An interesting development which will definitely open up new doors. 
I first came across MLIR in the [Tensorflow MLIR page](https://www.tensorflow.org/mlir). In case anyone is following this thread, I recommend you to have a look at the link.
